File: JSPTH/Math/Geometry.py
import math
from . import MainMath as MM

class GeometryMath():

    def sine(angle_rad: float) -> float:
        return (math.sin(angle_rad))

    def cosine(angle_rad: float) -> float:
        return(math.cos(angle_rad))



    def area_of_circle(radius: float) -> float:
        """
        Calculate the area of a circle.

        Parameters:
        radius (float): The radius of the circle.

        Returns:
        float: The area of the circle.
        """
        if radius < 0:
            raise ValueError("Radius cannot be negative")
        return(math.pi * radius**2)


    def circumference_of_circle(radius: float) -> float:
        """
        Calculate the circumference of a circle.

        Parameters:
        radius (float): The radius of the circle.

        Returns:
        float: The circumference of the circle.
        """
        if radius < 0:
            raise ValueError("Radius cannot be negative")
        return(2 * math.pi * radius)


    def volume_of_cone(radius: float, height: float) -> float:
        """
        Calculate the volume of a cone.

        Parameters:
        radius (float): The radius of the cone base.
        height (float): The height of the cone.

        Returns:
        float: The volume of the cone.
        """
        if radius < 0 or height < 0:
            raise ValueError("Radius or height cannot be negative")
        return((1/3) * math.pi * radius**2 * height)

    def surface_area_of_cone(radius: float, slant_height: float) -> float:
        """
        Calculate the surface area of a cone.

        Parameters:
        radius (float): The radius of the cone base.
        slant_height (float): The slant height of the cone.

        Returns:
        float: The surface area of the cone.
        """
        if radius < 0 or slant_height < 0:
            raise ValueError("Radius and slant height cannot be negative")
        return(math.pi * radius * (radius + slant_height))

    def distance_between_points(x1: float, y1: float, x2: float, y2: float) -> float:
        """
        Calculate the distance between two points in 2D space.

        Parameters:
        x1, y1 (float): Coordinates of the first point.
        x2, y2 (float): Coordinates of the second point.

        Returns:
        float: The distance between the two points.
        """
        return(math.sqrt((x2 - x1)**2 + (y2 - y1)**2))



    def triangle_area_heron(a: float, b: float, c: float) -> float:
        """
        Calculate the area of a triangle using Heron's formula.
        
        Parameters:
        a (float): Length of the first side.
        b (float): Length of the second side.
        c (float): Length of the third side.
        
        Returns:
        float: The area of the triangle.
        
        Raises:
        ValueError: If the sides do not form a valid triangle.
        """
        if a + b <= c or a + c <= b or b + c <= a:
            raise ValueError("The provided sides do not form a valid triangle.")
        
        # Semi-perimeter
        s = (a + b + c) / 2
        
        # Area calculation
        return math.sqrt(s * (s - a) * (s - b) * (s - c))
    


def circle_circumference(radius: float) -> float:
    """
    Calculate the circumference of a circle.
    
    Parameters:
    radius (float): The radius of the circle.
    
    Returns:
    float: The circumference of the circle.
    """
    return 2 * math.pi * radius


def circle_area(radius: float) -> float:
    """
    Calculate the area of a circle.
    
    Parameters:
    radius (float): The radius of the circle.
    
    Returns:
    float: The area of the circle.
    """
    return math.pi * radius ** 2


def parallelogram_area(base: float, height: float) -> float:
    """
    Calculate the area of a parallelogram.
    
    Parameters:
    base (float): The length of the base.
    height (float): The height of the parallelogram.
    
    Returns:
    float: The area of the parallelogram.
    """
    return base * height


def sphere_surface_area(radius: float) -> float:
    """
    Calculate the surface area of a sphere.
    
    Parameters:
    radius (float): The radius of the sphere.
    
    Returns:
    float: The surface area of the sphere.
    """
    return 4 * math.pi * radius ** 2

File: JavaScript/JSJS/MathJS/geometry.js
class GeometryUtils {
    constructor() {}
  
    // Area of a Circle: π * r²
    area_of_circle(radius) {
      if (radius < 0) throw new Error("Radius cannot be negative.");
      return Math.PI * radius ** 2;
    }
  
    // Circumference of a Circle: 2 * π * r
    circumference_of_circle(radius) {
      if (radius < 0) throw new Error("Radius cannot be negative.");
      return 2 * Math.PI * radius;
    }
  
    // Volume of a Cone: (1/3) * π * r² * h
    volume_of_cone(radius, height) {
      if (radius < 0 || height < 0) throw new Error("Radius and height cannot be negative.");
      return (1 / 3) * Math.PI * radius ** 2 * height;
    }
  
    // Surface Area of a Cone: π * r * (r + l), where l = slant height
    surface_area_of_cone(radius, height) {
      if (radius < 0 || height < 0) throw new Error("Radius and height cannot be negative.");
      const slantHeight = Math.sqrt(radius ** 2 + height ** 2);
      return Math.PI * radius * (radius + slantHeight);
    }
  
    // Distance Between Two Points: √((x2 - x1)² + (y2 - y1)²)
    distance_between_points(x1, y1, x2, y2) {
      return Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2);
    }
  
    // Triangle Area using Heron's Formula: √(s * (s - a) * (s - b) * (s - c)), where s = (a + b + c) / 2
    triangle_area_heron(a, b, c) {
      if (a < 0 || b < 0 || c < 0) throw new Error("Sides cannot be negative.");
      const s = (a + b + c) / 2;
      return Math.sqrt(s * (s - a) * (s - b) * (s - c));
    }
  
    // Circle Circumference: 2 * π * r
    circle_circumference(radius) {
      return this.circumference_of_circle(radius); // Reuses the existing function
    }
  
    // Circle Area: π * r²
    circle_area(radius) {
      return this.area_of_circle(radius); // Reuses the existing function
    }
  
    // Parallelogram Area: b * h
    parallelogram_area(base, height) {
      if (base < 0 || height < 0) throw new Error("Base and height cannot be negative.");
      return base * height;
    }
  
    // Surface Area of a Sphere: 4 * π * r²
    sphere_surface_area(radius) {
      if (radius < 0) throw new Error("Radius cannot be negative.");
      return 4 * Math.PI * radius ** 2;
    }
  
    // Volume of a Sphere: (4/3) * π * r³
    sphere_volume(radius) {
      if (radius < 0) throw new Error("Radius cannot be negative.");
      return (4 / 3) * Math.PI * radius ** 3;
    }
  }
  